jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -34,183 +34,238 @@
+                    Preconditions.checkNotNull(steps.pushContainerConfigurationStep),
+                    Preconditions.checkNotNull(steps.buildImageStep)));
   }
 
-  public StepsRunner runWriteTarFileStep(Path outputPath) {
-    writeTarFileStep =
-        new WriteTarFileStep(
-            listeningExecutorService,
-            outputPath,
-            buildConfiguration,
-            Preconditions.checkNotNull(pullAndCacheBaseImageLayersStep),
-            Preconditions.checkNotNull(buildAndCacheApplicationLayerSteps),
-            Preconditions.checkNotNull(buildImageStep));
-    return this;
+  public StepsRunner loadDocker(DockerClient dockerClient) {
+    return enqueueStep(
+        () ->
+            steps.loadDockerStep =
+                new LoadDockerStep(
+                    listeningExecutorService,
+                    dockerClient,
+                    buildConfiguration,
+                    Preconditions.checkNotNull(steps.pullAndCacheBaseImageLayersStep),
+                    Preconditions.checkNotNull(steps.buildAndCacheApplicationLayerSteps),
+                    Preconditions.checkNotNull(steps.buildImageStep)));
+  }
+
+  public StepsRunner writeTarFile(Path outputPath) {
+    return enqueueStep(
+        () ->
+            steps.writeTarFileStep =
+                new WriteTarFileStep(
+                    listeningExecutorService,
+                    outputPath,
+                    buildConfiguration,
+                    Preconditions.checkNotNull(steps.pullAndCacheBaseImageLayersStep),
+                    Preconditions.checkNotNull(steps.buildAndCacheApplicationLayerSteps),
+                    Preconditions.checkNotNull(steps.buildImageStep)));
+  }
+
+  public BuildResult waitOnPushImage() throws ExecutionException, InterruptedException {
+    stepsRunnable.run();
+    return Preconditions.checkNotNull(steps.pushImageStep).getFuture().get();
   }
 
-  public BuildResult waitOnPushImageStep() throws ExecutionException, InterruptedException {
-    return Preconditions.checkNotNull(pushImageStep).getFuture().get();
+  public BuildResult waitOnLoadDocker() throws ExecutionException, InterruptedException {
+    stepsRunnable.run();
+    return Preconditions.checkNotNull(steps.loadDockerStep).getFuture().get();
   }
 
-  public BuildResult waitOnLoadDockerStep() throws ExecutionException, InterruptedException {
-    return Preconditions.checkNotNull(loadDockerStep).getFuture().get();
+  public BuildResult waitOnWriteTarFile() throws ExecutionException, InterruptedException {
+    stepsRunnable.run();
+    return Preconditions.checkNotNull(steps.writeTarFileStep).getFuture().get();
   }
 
-  public BuildResult waitOnWriteTarFileStep() throws ExecutionException, InterruptedException {
-    return Preconditions.checkNotNull(writeTarFileStep).getFuture().get();
+  private StepsRunner enqueueStep(Runnable stepRunnable) {
+    stepsRunnable =
+        () -> {
+          stepsRunnable.run();
```

Comment:
Hmm... not sure if this will work, because it's defined recursively using itself. Needs verification.

Reply:
Ah right, needs to finalize in another variable.